service/history/shard/context_impl.go

```diff
@@ -1203,18 +1207,23 @@ func (s *ContextImpl) updateShardInfo(
 		return err
 	}
 
+	s.tasksCompletedSinceLastUpdate += tasksCompleted
 	updateFnLocked()
 	s.shardInfo.StolenSinceRenew = 0
 
 	now := s.timeSource.Now()
-	if s.lastUpdated.Add(s.config.ShardUpdateMinInterval()).After(now) {
+	notFirstUpdate := !s.lastUpdated.IsZero()
+	notEnoughTasksCompleted := s.tasksCompletedSinceLastUpdate < s.config.ShardUpdateMinTasksCompleted()
```

Comment:
I think we probably need some more metrics to help us tune the `ShardUpdateMinTasksCompleted` config. 
e.g. metric for the # of tasks completed. or the duration between two shard info update.

Reply:
How about two histograms: "tasks completed per update" and "time between updates"

Reply:
Both have been added in my latest commit